### PR TITLE
Problem: cosmos-sdk 0.46 is not used

### PR DESCRIPTION
**Pull request by yihuang**
👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

- [x] fix build
- [x] fix go lint
- [x] fix unit tests
- [x] fix integration tests
- [x] fix sim tests

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: app/test_helpers.go [hunk 1]
```diff
@@ -380,7 +382,8 @@ func GenSequenceOfTxs(txGen client.TxConfig, msgs []sdk.Msg, accNums []uint64, i
 	txs := make([]sdk.Tx, numToGenerate)
 	var err error
 	for i := 0; i < numToGenerate; i++ {
-		txs[i], err = helpers.GenTx(
+		txs[i], err = helpers.GenSignedMockTx(
+			rand.New(rand.NewSource(1)),
```
[the remark this answers is not in the document]
Comment (yihuang) [reply app/test_helpers.go:387]: @sonatype-lift ignore

File: app/test_helpers.go [hunk 2]
```diff
@@ -330,7 +331,8 @@ func SignCheckDeliver(
 	chainID string, accNums, accSeqs []uint64, expSimPass, expPass bool, priv ...cryptotypes.PrivKey,
 ) (sdk.GasInfo, *sdk.Result, error) {
 
-	tx, err := helpers.GenTx(
+	tx, err := helpers.GenSignedMockTx(
+		rand.New(rand.NewSource(1)),
```
[the remark this answers is not in the document]
Comment (yihuang) [reply app/test_helpers.go:336]: @sonatype-lift ignore

File: x/chainmain/client/cli/query.go [hunk 3]
```diff
@@ -10,12 +10,12 @@ import (
 	"github.com/cosmos/cosmos-sdk/client"
 	"github.com/cosmos/cosmos-sdk/client/flags"
 	sdk "github.com/cosmos/cosmos-sdk/types"
-	"github.com/cosmos/cosmos-sdk/types/rest"
+	"github.com/cosmos/cosmos-sdk/types/query"
```
[the remark this answers is not in the document]
Comment (yihuang) [reply x/chainmain/client/cli/query.go:13]: @sonatype-lift ignore

File: x/nft/types/keys.go [hunk 4]
```diff
@@ -13,6 +13,9 @@ const (
 	// ModuleName is the name of the module
 	ModuleName = "nft"
 
+	// avoid conflicts with sdk nft module
+	ModuleNameAlias = "nft2"
+
 	// StoreKey is the default store key for NFT
 	StoreKey = ModuleName
```
Comment (tomtau) [inline x/nft/types/keys.go:20]: wouldn't there be a conflict on the store?

Comment (yihuang) [reply x/nft/types/keys.go:20]: as long as we don't enable the SDK module, it might be fine?

File: .golangci.yml [hunk 5]
```diff
@@ -17,6 +17,12 @@ linters:
     - stylecheck
     - unconvert
 
+issues:
+  exclude-rules:
+    - text: "Use of weak random number generator"
+      linters:
+        - gosec
```
[the remark this answers is not in the document]
Comment (yihuang) [reply .golangci.yml]: done

File: x/nft/types/keys.go (hunk 4 above)
Comment (tomtau) [reply x/nft/types/keys.go:20]: but we will want to migrate to it / enable it?

Comment (tomtau) [reply x/nft/types/keys.go:20]: https://github.com/crypto-org-chain/chain-main/blob/master/doc/architecture/adr-004.md

Comment (yihuang) [reply x/nft/types/keys.go:20]: hmm, maybe do that in another PR: https://github.com/crypto-org-chain/chain-main/issues/829

**Comment by devashishdxt** [comment]
I was hoping to do these changes in multiple steps (so that code review is easier):

1. Remove `wasmd`
2. Change release scripts to not include dlls
3. Create an `alpha` release with `cosmos-sdk 0.45.7` and `ibc-go 4.0.0` (for croeseid-5)
4. Upgrade `cosmos-sdk` to `v0.46.0` and `ibc-go` to `v5.0.0` (without enabling native NFT module)
5. Migrate from custom NFT module to native NFT module

@yihuang @tomtau Can we wait to merge this until next `alpha` release? Otherwise, we'll have to do duplicate work on `release/v4` branch.

**Comment by tomtau** [comment]
@devashishdxt fine to break it down in that sequence, but is there a need to wait for `ibc-go 4.0.0`?

**Comment by devashishdxt** [comment]
> @devashishdxt fine to break it down in that sequence, but is there a need to wait for `ibc-go 4.0.0`? or is rc2 ok?

Final version will use SDK `v0.45.7` (which has fixes for non-determinism in simulation tests).

**Comment by tomtau** [comment]
> > @devashishdxt fine to break it down in that sequence, but is there a need to wait for `ibc-go 4.0.0`? or is rc2 ok?
> 
> Final version will use SDK `v0.45.7` (which has fixes for non-determinism in simulation tests).

that could be achieved with `replace` or a fork of ibc-go; no need to wait for the final release here (given it's for the alpha testnet?)

**Comment by devashishdxt** [comment]
> that could be achieved with `replace` or a fork of ibc-go; no need to wait for the final release here (given it's for the alpha testnet?)

I've talked to the team, they're going to release rc3 today. We can use that instead of using replace/fork.

**Comment by tomtau** [comment]
@devashishdxt for 
> 2. Change release scripts to not include dlls
are you going to submit a PR or waiting for @yihuang ?

**Comment by devashishdxt** [comment]
> 2\. Change release scripts to not include dlls

Hi @yihuang, can you help me create a PR for this?

**Comment by yihuang** [comment]
> > 2. Change release scripts to not include dlls
> 
> Hi @yihuang, can you help me create a PR for this?

revert to old goreleaser workflow, right?

**Comment by devashishdxt** [comment]
> > > 2. Change release scripts to not include dlls
> > 
> > 
> > Hi @yihuang, can you help me create a PR for this?
> 
> revert to old goreleaser workflow, right?

Yes. The PR has other changes too, so, I'm not sure which ones to keen and which ones to revert.

**Comment by yihuang** [comment]
> > > > 2. Change release scripts to not include dlls
> > > 
> > > 
> > > Hi @yihuang, can you help me create a PR for this?
> > 
> > 
> > revert to old goreleaser workflow, right?
> 
> Yes. The PR has other changes too, so, I'm not sure which ones to keen and which ones to revert.

sure, https://github.com/crypto-org-chain/chain-main/pull/832

**Comment by tomtau** [comment]
> 3. Create an alpha release with cosmos-sdk 0.45.7 and ibc-go 4.0.0 (for croeseid-5)
@devashishdxt are you going to tag this one?

**Comment by devashishdxt** [comment]
> > 3. Create an alpha release with cosmos-sdk 0.45.7 and ibc-go 4.0.0 (for croeseid-5)
> 
> 
> 
> @devashishdxt are you going to tag this one?

Yes. Just running a few things locally. Will create a PR then for `release/...` branch.

File: go.mod [hunk 6]
```diff
@@ -107,38 +127,43 @@ require (
...
 
 	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
 
-	// REF: https://github.com/cosmos/cosmos-sdk/blob/main/docs/basics/app-anatomy.md#dependencies-and-makefile
-	google.golang.org/grpc => google.golang.org/grpc v1.33.2
+	// TODO: remove after fixed https://github.com/cosmos/cosmos-sdk/issues/11364
+	github.com/zondax/hid => github.com/zondax/hid v0.9.0
 )
+
+replace github.com/cosmos/cosmos-sdk => github.com/yihuang/cosmos-sdk v0.43.0-beta1.0.20220811091413-ef8c1f4e5d33
```
Comment (yihuang) [inline go.mod:169]: https://github.com/cosmos/cosmos-sdk/pull/12906

Comment (tomtau) [reply go.mod]: it looks this was backported: https://github.com/cosmos/cosmos-sdk/tree/release/v0.46.x ?

Comment (yihuang) [reply go.mod]: yes, dependency updated.

File: app/app.go [hunk 7]
```diff
@@ -19,7 +19,6 @@ import (
 	"github.com/cosmos/cosmos-sdk/baseapp"
```
Comment (tomtau) [inline app/app.go:19]: should x/group (https://github.com/cosmos/cosmos-sdk/releases/tag/v0.46.0) be added in this PR or a separate PR?

File: go.mod [hunk 8]
```diff
@@ -107,38 +127,42 @@ require (
...
 	golang.org/x/term v0.0.0-20220526004731-065cf7ba2467 // indirect
 	golang.org/x/text v0.3.7 // indirect
+	golang.org/x/xerrors v0.0.0-20220517211312-f3a8303e98df // indirect
+	google.golang.org/api v0.81.0 // indirect
+	google.golang.org/appengine v1.6.7 // indirect
 	gopkg.in/ini.v1 v1.66.6 // indirect
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 	nhooyr.io/websocket v1.8.6 // indirect
+	sigs.k8s.io/yaml v1.3.0 // indirect
 )
```
Comment (tomtau) [inline go.mod:155]: good to check if some of the replace entries are still necessary, such as github.com/cosmos/ledger-cosmos-go

File: integration_tests/test_gov.py [hunk 9]
```diff
@@ -18,7 +18,7 @@ def test_param_proposal(cluster, vote_option):
     """
```
Comment (tomtau) [inline integration_tests/test_gov.py:18]: one TODO to have tests for the new gov proposals?

**Review by tomtau — APPROVED** [review]
I guess x/group could be added and tested in a separate PR?

File: integration_tests/test_gov.py (hunk 9 above)
Comment (yihuang) [reply integration_tests/test_gov.py:18]: opened: https://github.com/crypto-org-chain/chain-main/issues/835

File: app/app.go (hunk 7 above)
Comment (yihuang) [reply app/app.go:19]: maybe a different PR with corresponding integration test together?

File: go.mod (hunk 8 above)
Comment (yihuang) [reply go.mod:155]: it seems the ledger-cosmos-go didn't make a new release with that.

File: app/app.go (hunk 7 above)
Comment (tomtau) [reply app/app.go:19]: yes + upgrade handler

File: go.mod (hunk 8 above)
Comment (yihuang) [reply go.mod:155]: the other replacements look reasonable.

**Comment by tomtau** [comment]
@yihuang can you rebase on the latest branch?

**Review by tomtau — COMMENTED** [review]
@devashishdxt should this PR also add `x/group` or should that be done separately?

**Comment by devashishdxt** [comment]
> @devashishdxt should this PR also add `x/group` or should that be done separately?

It's better to do it separately. This PR should just upgrade the cosmos-sdk version without adding any extra functionality. New things in cosmos-sdk can be adding in separate PRs.

**Comment by tomtau** [comment]
https://github.com/crypto-org-chain/chain-main/pull/828#issuecomment-1209296723
so are you going to do it before the NFT module work?

**Comment by devashishdxt** [comment]
> https://github.com/crypto-org-chain/chain-main/pull/828#issuecomment-1209296723
> 
> so are you going to do it before the NFT module work?

Yes

**Comment by yihuang** [comment]
> @yihuang can you rebase on the latest branch?

conflicts fixed.

**Comment by yihuang** [comment]
the timeout change seems don't have effect

**Comment by tomtau** [comment]
maybe the timeout change may need to be merged to the upstream branch first?

**Comment by yihuang** [comment]
> maybe the timeout change may need to be merged to the upstream branch first?

maybe, I'm running locally.

**Comment by tomtau** [comment]
https://github.com/crypto-org-chain/chain-main/pull/839

**Comment by yihuang** [comment]
all tests passed, ready to be merged anytime @devashishdxt

File: app/app.go [hunk 10]
```diff
@@ -587,23 +594,11 @@ func New(
...
-		slashing.NewAppModule(appCodec, app.SlashingKeeper, app.AccountKeeper, app.BankKeeper, app.StakingKeeper),
-		params.NewAppModule(app.ParamsKeeper),
-		evidence.NewAppModule(app.EvidenceKeeper),
-		authzmodule.NewAppModule(appCodec, app.AuthzKeeper, app.AccountKeeper, app.BankKeeper, app.interfaceRegistry),
-		ibc.NewAppModule(app.IBCKeeper),
-		transferModule,
-		nft.NewAppModule(appCodec, app.NFTKeeper, app.AccountKeeper, app.BankKeeper),
-	)
+	overrideModules := map[string]module.AppModuleSimulation{
+		// Use custom RandomGenesisAccounts so that auth module could create random EthAccounts in genesis state when genesis.json not specified
```
Comment (tomtau) [inline app/app.go:598]: I guess this comment isn't relevant here?

**Review by tomtau — APPROVED** [review]
there are conflicts with the latest branch, can you rebase it?

**Comment by tomtau** [comment]
oh, squash and merge should work
